src/admin/metrics.rs

```diff
@@ -0,0 +1,336 @@
+use http_body_util::Full;
+use hyper::{
+    body::Bytes,
+    server::conn::http1,
+    service::service_fn,
+    Request,
+};
+use prometheus_metric_storage::{
+    MetricStorage,
+    StorageRegistry,
+};
+use std::convert::Infallible;
+use std::{
+    collections::hash_map::HashMap,
+    sync::{
+        Arc,
+        RwLock,
+    },
+    time::Duration,
+};
+
+use tokio::sync::{
+    mpsc::{
+        unbounded_channel,
+        UnboundedReceiver,
+        UnboundedSender,
+    },
+    oneshot,
+};
+
+use thiserror::Error;
+
+pub type MetricSender = UnboundedSender<RpcMetrics>;
+pub type MetricReceiver = UnboundedReceiver<RpcMetrics>;
+
+// #[cfg(feature = "prometheusd")]
+#[derive(Debug, Error)]
+pub enum MetricsError {
+    #[error(transparent)]
+    WSError(#[from] crate::websocket::error::WsError),
+    #[error(transparent)]
+    RPCError(#[from] crate::rpc::error::RpcError),
+}
+
+// #[cfg(feature = "prometheusd")]
+impl From<MetricsError> for String {
+    fn from(error: MetricsError) -> Self {
+        error.to_string()
+    }
+}
+#[derive(MetricStorage, Clone, Debug)]
+#[metric(subsystem = "rpc")]
+pub struct RpcMetrics {
+    #[metric(labels("path", "method", "status"), help = "Total number of requests")]
+    pub requests: prometheus::IntCounterVec,
+    #[metric(labels("path", "method"), help = "latency of request")]
+    pub duration: prometheus::HistogramVec,
+}
+impl RpcMetrics {
+    pub fn init(registry: &StorageRegistry) -> Result<&Self, prometheus::Error> {
+        RpcMetrics::instance(registry)
+    }
+    pub fn requests_complete(&self, path: &str, method: &str, status: &u16, dt: Duration) {
+        self.requests
+            .with_label_values(&[path, method, &status.to_string()])
+            .inc();
+        self.duration
+            .with_label_values(&[path, method])
+            .observe(dt.as_millis() as f64)
+    }
+}
+#[cfg(feature = "prometheusd")]
+#[derive(Debug)]
+pub enum MetricsCommand<'a> {
+    Flush(),
+    Channel(&'a MetricSender, &'a MetricReceiver),
+    PushLatency(&'a RpcMetrics, &'a str, &'a str, f64),
+    PushRequest(&'a RpcMetrics, &'a str, &'a str, &'a u16, Duration),
+    PushError(&'a RpcMetrics, &'a str, &'a str, &'a u16, Duration),
+    Push(&'a RpcMetrics),
+    Pull(&'a RpcMetrics),
+}
+
+#[cfg(feature = "prometheusd")]
+#[derive(Debug)]
+pub enum MetricChannelCommand {
+    AdminMsg(oneshot::Sender<RpcMetrics>),
+    StatsMsg(oneshot::Sender<RpcMetrics>),
+}
+
+#[cfg(not(feature = "prometheusd"))]
+pub async fn metrics_update_sink(mut metrics_rx: MetricReceiver) {
+    loop {
+        while metrics_rx.recv().await.is_some() {
+            continue;
+        }
+    }
+}
+
+#[cfg(feature = "prometheusd")]
+pub async fn listen_for_metrics_requests(
+    metrics_rx: MetricReceiver,
+    registry_status: Arc<RwLock<StorageRegistry>>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let address = "127.0.0.1:9091";
```

Comment:
obv this shouldnt be hardocoded, just be aware about how youre planning to get it from the config!